x/mongo/driver/topology/server_test.go

```diff
@@ -309,6 +312,89 @@ func TestServer(t *testing.T) {
+
+		sdam := &event.ServerMonitor{
+			ServerHeartbeatStarted:   serverHeartbeatStarted,
+			ServerHeartbeatSucceeded: serverHeartbeatSucceeded,
+			ServerHeartbeatFailed:    serverHeartbeatFailed,
+		}
+
+		dialer := &channelNetConnDialer{}
+		dialerOpt := WithDialer(func(Dialer) Dialer {
+			return dialer
+		})
+		serverOpts := []ServerOption{
+			WithConnectionOptions(func(connOpts ...ConnectionOption) []ConnectionOption {
+				return append(connOpts, dialerOpt)
+			}),
+			withMonitoringDisabled(func(bool) bool { return true }),
+			WithServerMonitor(func(*event.ServerMonitor) *event.ServerMonitor { return sdam }),
+		}
+
+		s, err := NewServer(address.Address("localhost:27017"), primitive.NewObjectID(), serverOpts...)
+		if err != nil {
+			t.Fatalf("error from NewServer: %v", err)
+		}
+
+		// set up heartbeat connection, which doesn't send events
+		_, err = s.check()
+		assert.Nil(t, err, "check error: %v", err)
+
+		channelConn := s.conn.nc.(*drivertest.ChannelNetConn)
+		_ = channelConn.GetWrittenMessage()
+
+		t.Run("success", func(t *testing.T) {
+			publishedEvents = nil
+			// do a heartbeat with a non-nil connection
+			if err = channelConn.AddResponse(makeIsMasterReply()); err != nil {
+				t.Fatalf("error adding response: %v", err)
+			}
+			_, err = s.check()
+			assert.Nil(t, err, "check error: %v", err)
+
+			assert.Equal(t, len(publishedEvents), 2, "expected %v events, got %v", 2, len(publishedEvents))
+
+			started, ok := publishedEvents[0].(event.ServerHeartbeatStartedEvent)
+			assert.True(t, ok, "expected type %T, got %T", event.ServerHeartbeatStartedEvent{}, publishedEvents[0])
+			assert.Equal(t, started.ConnectionID, s.conn.ID(), "expected connectionID to match")
+
+			succeeded, ok := publishedEvents[1].(event.ServerHeartbeatSucceededEvent)
+			assert.True(t, ok, "expected type %T, got %T", event.ServerHeartbeatSucceededEvent{}, publishedEvents[1])
+			assert.Equal(t, succeeded.ConnectionID, s.conn.ID(), "expected connectionID to match")
+			assert.Equal(t, succeeded.Reply.Addr, s.address, "expected address %T, got %T", s.address, succeeded.Reply.Addr)
+		})
+		t.Run("failure", func(t *testing.T) {
+			publishedEvents = nil
+			// do a heartbeat with a non-nil connection
+			channelConn.ReadErr <- errors.New("error")
+			_, err = s.check()
+			assert.Nil(t, err, "check error: %v", err)
+
+			assert.Equal(t, len(publishedEvents), 2, "expected %v events, got %v", 2, len(publishedEvents))
+
+			started, ok := publishedEvents[0].(event.ServerHeartbeatStartedEvent)
+			assert.True(t, ok, "expected type %T, got %T", event.ServerHeartbeatStartedEvent{}, publishedEvents[0])
+			assert.Equal(t, started.ConnectionID, s.conn.ID(), "expected connectionID to match")
+
+			failed, ok := publishedEvents[1].(event.ServerHeartbeatFailedEvent)
+			assert.True(t, ok, "expected type %T, got %T", event.ServerHeartbeatFailedEvent{}, publishedEvents[1])
+			assert.Equal(t, failed.ConnectionID, s.conn.ID(), "expected connectionID to match")
```

Comment:
Assert `failed.Awaited` is false and that `failed.Failure` is equal to the expected error (if that's reasonable).